File: 2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.go
func repairCars(ranks []int, cars int) int64 {
	cars64 := int64(cars);

	left := int64(0);
	right := int64(ranks[0]) * cars64 * cars64;

	for left < right {
		mid := left + (right - left) / 2;
		totalCars := int64(0);

		for _, r := range ranks {
			n := int64(math.Sqrt(float64(mid) / float64(r)));
			totalCars += n;
		}

		if totalCars >= cars64 {
			right = mid;
		} else {
			left = mid + 1;
		}
	}

	return left;
}

File: 2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.cpp
class Solution {
public:
    inline bool isPossible(long long int time, int cars, vector<int>& ranks) {
        for (int& it : ranks) {
            long long int sq_val = time / (long long int)it;
            long long int n = sqrt(sq_val);
            
            cars -= n;
            if (cars <= 0)
                return true;
        }

        return cars <= 0;
    }

    long long repairCars(vector<int>& ranks, int cars) {
        long long int start = 0, end = cars;
        end *= cars;
        end *= *max_element(ranks.begin(), ranks.end());

        long long int mid, ans = end;
        while (start <= end) {
            mid = (end - start) / 2 + start;
            if (isPossible(mid, cars, ranks)) {
                ans = mid;
                end = mid - 1;
            } else
                start = mid + 1;
        }

        return ans;
    }
};